src/pymor/algorithms/basic.py

```diff
@@ -67,3 +67,37 @@ def almost_equal(U, V, product=None, norm=None, rtol=1e-14, atol=1e-14):
     ERR_norm = norm(X)
 
     return ERR_norm <= atol + V_norm * rtol
+
+
+def relative_error(U, V, product=None):
+    """Compute error between U and V relative to norm of U."""
+    return (U - V).norm(product) / U.norm(product)
+
+
+def project_array(U, basis, product=None, orthonormal=True):
+    """Orthogonal projection of |VectorArray| onto subspace.
+
+    Parameters
+    ----------
+    U
+        The |VectorArray| to project.
+    basis
+        |VectorArray| of basis vectors for the subspace onto which
+        to project.
+    product
+        Inner product |Operator| w.r.t. which to project.
+    orthonormal
+        If `True`, the vectors in `basis` are assumed to be orthonormal
+        w.r.t. `product`.
+
+    Returns
+    -------
+    The projected |VectorArray|.
+    """
+    if orthonormal:
+        return basis.lincomb(U.inner(basis, product))
+    else:
+        gramian = basis.gramian(product)
+        rhs = basis.inner(U, product)
+        coeffs = np.linalg.solve(gramian, rhs).T
+        return basis.lincomb(coeffs)
```

Comment:
Is there a reason why not use `basis = gram_schmidt(basis, product=product, atol=0, rtol=0)` here? I would guess `gramian` could be ill-conditioned.

Reply:
Good point! A problem with `gram_schmidt` might be that one would have to make a copy of `basis`, which potentially requires a lot of memory. Not sure what to prefer ...

Reply:
The user could always call `gram_schmidt` outside and then set `orthonormal` to `True`. How about keeping the current code and additionally computing the condition of `gramian`? Then one could issue a warning, if the condition is too large. - On the other hand, the condition probably gets large very quickly. Any opinions, @pmli?

Reply:
@sdrave How about adding `copy=False` to the `gram_schmidt` call? (That would actually make more sense than my first suggestion.)
(BTW, since recently, [`scipy.linalg.solve` raises a warning if the system is ill-conditioned](https://github.com/scipy/scipy/blob/a8229ce213ea7731afb4a4a91e4681a0821ba40b/scipy/linalg/basic.py#L107-L108).)

Reply:
`copy=False` is not really an option. The caller really would not expect `basis` to change, just by computing a projection.

Reply:
Aha, ok. Then replacing `numpy.linalg.solve` by `scipy.linalg.solve` would be good.